### ISSUE TITLE
Don't require this as the default mesos ui

### ISSUE THREAD

**Issue by SEJeff** [Capgemini/mesos-ui#57]
Say I want to keep the default mesos ui, and present this as the ui I show users via a service discovery solution or something.

Would it be possible to use this and have it reach out to the mesos master via the http api?

**Comment by enxebre** [comment]
Hey SEJeff, we have removed the dependency with the nodejs server for the sake of a better integration with mesos and zookeeper and to line up with mesos default mechanism.

You can still run the standalone version which fits that use case by running https://github.com/Capgemini/mesos-ui/tree/0.1.1

You can see latest changes here https://github.com/Capgemini/mesos-ui/blob/master/CHANGELOG.md

**Comment by SEJeff** [comment]
So basically, there is no way to run the latest version of this without requiring it to be the default mesos ui? Hmmmm. That is a real bummer. Is there any hope to have this merged upstream as the default mesos ui longer term?

I realize I could use the older version, but that isn't really what I was going for.

**Comment by wallies** [comment]
hey @SEJeff, so i think you may have two options, depending how you are running mesos. So you could either run gulp build command which will give you the static files and then run them as a static node server using NODE_ENV===prod node build/server.js or you could build a container that does much the same thing. I have created #58 to deal with this.

**Comment by SEJeff** [comment]
@wallies That seems perfect. Then that can be executed via marathon. Thanks!

**Comment by enxebre** [comment]
@wallies, @SEJeff You can just modify the mesosEndpoint in the config file to point to the Mesos server  and just run "gulp" which will run a nodejs basic server serving the mesos-ui talking to the mesos server but the way this is at the moment it only would work if both the mesos-ui and mesos share the same domain otherwise It will fail due to "Same-origin_policy" unless we have a solution for allowing CORS somehow.
@SEJeff we decided to choose just one for the sake of code simplicity and we went for this one as is it is more Mesos core friendly.

It'd be great if we can support both in a simple way similar to what @wallies suggests

**Comment by SEJeff** [comment]
@enxebre makes sense. Being a python guy, I whipped up this quick flask proxy to add the access control:

```python
import os
import urllib
import urlparse
from flask import Flask
from flask import request
from flask import Response
from flask import stream_with_context

import requests

app = Flask(__name__)
MESOS_ENDPOINT = os.environ.get('MESOS_ENDPOINT', '<hardcoded default here>')

@app.route('/<path:url>')
def home(url):
    args = urllib.urlencode(request.args.items())
    url = '{url}?{args}'.format(url=url, args=args)
    url = urlparse.urljoin(MESOS_HOST, url)
    req = requests.get(url, stream=True)

    return Response(
        stream_with_context(req.iter_content()),
        content_type=req.headers['content-type'],
        headers={'Access-Control-Allow-Origin': '*'},
        status=req.status_code,
    )

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5050, debug=True)
```

Any idea if there is a mesos issue for allowing configuration of CORS? I'm happy to open a jira upstream for that as it seems useful for tons of projects (like this one).

**Comment by wallies** [comment]
@SEJeff I am going to add this cors proxy (https://www.npmjs.com/package/corsproxy) into the server.js, which should help things work for cors in the short term. I think we need to raise a ticket upstream to get CORS configured for mesos.

**Comment by SEJeff** [comment]
@wallies and @enxebre I really appreciate it. I'll go file a Jira for mesos now.

**Comment by enxebre** [comment]
:+1:

**Comment by SEJeff** [comment]
[MESOS-3796](https://issues.apache.org/jira/browse/MESOS-3796) has been filed!

**Comment by enxebre** [comment]
awesome! just in case you didn't know we've tagged the docker images so you could use 0.1.1 for your use case in the meanwhile.
https://hub.docker.com/r/capgemini/mesos-ui/tags/

**Comment by WalterDalton** [comment]
@enxebre  I am interested in the logic on only supporting the one code base. While I agree it's easier and cleaner to support one, I think that should have been a goal first, and then ensured all things that we use the Mesos default UI were supported in your project. Basically, I look at how I use the default-UI, the first thing I see missing in mesos-ui is a running task list.  (Granted I have not gotten a recent version running, I don't want to replace the default-ui if I don't have the features I need, and the screen shots don't show this yet...) The task list is huge for me, it helps me troubleshoot and drill down into the individual sandboxes and figure out failures etc.  Without this, I can't troubleshoot things easily, so I need this functionality. But if I replace default-ui with mesos-ui, now I have to give that up. 

I guess what I am saying is I love your project. I ran it in an previous form without the integration, I went to go install a new version and was discouraged by the default-ui replacement. (I understand there is another branch I can go to, but it's not updated... so am I missing stuff? etc).  I guess, I would have liked to see both be first class citizens of a project like this until such time the default-ui could be completely replaced. 

Now, I know I am coming off as the "the internet is complaining again", that's not my intent. I love what you have done with this project, and I am trying to help provide some use cases that work against the the logic of default-ui replacement only (in it's current state).  In addition, if I am missing something (if the ability to see tasks, and drill into sandboxes has been added) by posting this, I hope to get a helpful "it's already in it" pat on the head.  

Thanks for all your hard work.

**Comment by enxebre** [comment]
Hey @WalterDalton all you are saying makes a lot of sense to me :)
As I said before we removed the dependency with the nodejs server for a better integration with mesos and zookeeper and to line up with mesos default mechanism and also because we got some feedback from the mesos community asking for this feature. By the look of the number of downloads on dockerhub seems like this approach is being quite useful(although to be fair I wouldn't be surprise if there's an automated script some where..)
But @SEJeff, @WalterDalton I completely see your point, as you said it especially true while this UI is still missing essential features, so this should provide support for running the app either on marathon or as Mesos default ui while sharing same code base. https://github.com/Capgemini/mesos-ui/pull/63
So then we can publish 2 docker images every release.

I completely agree about the importance of the task list component. One of the ideas behind this project was to get people from the community involved and make the most of reactjs modularisation via components  in order to inflate the features of the ui in a collaborative way based on the community demand.
We'll be working on a "task list" component and other features considered indispensable but unfortunately the commit rate won't be as fast as we'd like as things happen at work as they do.

Many thanks for your comments and feedback!
